torch/_inductor/fx_passes/post_grad.py

```diff
@@ -627,44 +630,69 @@ def reinplace_scatters(graph):
     for node in reversed(graph.nodes):
         storage_to_nodes[get_node_storage(node)].append(node)
         if node.target == aten.copy_.default:
-            copy_args_to_copy_nodes[(node.args[0], node.args[1])] = node
+            src = node.args[0]
+            dst = node.args[1]
+            # If the target is a getitem and it indexes a possible clone,
+            # then skip over it
+            if (
+                dst.target == operator.getitem
+                and dst.args[0].kwargs["kwargs"][dst.args[1]] == node.args[0]
+            ):
+                dst = dst.args[0]
+            copy_args_to_copy_nodes[(src, dst)] = node
             assert node.args[0].op == "placeholder"
             mutated_inputs.add(node.args[0])
 
+    def can_replace(node, mutated_arg):
+        if get_node_storage(mutated_arg) is None:
+            return False
+        shared_view_nodes = storage_to_nodes[get_node_storage(mutated_arg)]
+        if mutated_arg.op == "placeholder":
+            if not (
+                copy_node := copy_args_to_copy_nodes.get((mutated_arg, node), False)
+            ):
+                return False
+
+            if len(shared_view_nodes) > 2:  # Arg aliases another node other than copy_
+                return False
+
+            # Check for any uses other than current node and copy_ epilogue
+            if len(mutated_arg.users) > 2:
```

Comment:
I think this check will miss some cases. In particular, `mutated_arg` can have more than 2 users, but it can still be ok to reinplace. Example:
```
def f(x, out):
    # x will now have **three** users: `add`, `triton_kernel`, and `copy_`
    tmp = torch.add(x, 1)
    triton_kernel[grid](inp=x, out=out)
    return out, tmp
```

Hmm, I think what we probably want is: if we look at users of `mutated_arg` **later** in the graph than the triton kernel, `copy_()` should be the only user (if this is the case, then it's safe to reinplace).

I'm not sure if there's a builtin way in FX to check "number of users after a given node", but there's a util here that does something similar: https://github.com/pytorch/pytorch/blob/main/torch/fx/passes/reinplace.py#L176

cc @Chillee (lmk if this sounds right to you)

Reply:
Yeah, @Chillee and I talked about this. As you mentioned, this could be improved by doing exactly what you suggested.

Reply:
I think the concrete check is

"are there any views of this node that are used after the mutation"